test/dataloader2/test_dataloader2.py

```diff
@@ -628,6 +619,42 @@ def test_non_replicable_datapipe(self, ctx) -> None:
         torch.manual_seed(321)
         self.assertNotEqual(res, list(dl) + list(dl))
 
+    @parametrize("num_workers", [1, 3])
+    @parametrize("worker_prefetch_cnt", [0, 5, 10])
+    def test_worker_exception_raised(self, num_workers, worker_prefetch_cnt):
+        dp = IterableWrapper(range(100)).sharding_filter()
+        dp = MakeMistakeDataPipe(dp)
+        rs = MultiProcessingReadingService(num_workers=num_workers, worker_prefetch_cnt=worker_prefetch_cnt)
+        dl = DataLoader2(dp, reading_service=rs)
+        it = iter(dl)
+        for _ in range(EXCEPTION_ITERATION_NUM * num_workers):
+            next(it)
+        with self.assertRaises(_CustomException) as cm:
+            next(it)
+        exc_msg = str(cm.exception)
+        self.assertTrue("Caught _CustomException in worker process 0" in exc_msg)
+        self.assertTrue("Original Traceback" in exc_msg)
+        self.assertTrue("_CustomException: oops" in exc_msg)
+
+    @parametrize("num_workers", [1, 3])
+    @parametrize("worker_prefetch_cnt", [0, 5, 10])
+    def test_dispatching_exception_raised(self, num_workers, worker_prefetch_cnt):
+        dp = IterableWrapper(range(100))
+        dp = MakeMistakeDataPipe(dp)
+        dp = dp.sharding_round_robin_dispatch(SHARDING_PRIORITIES.MULTIPROCESSING)
+        dp = dp.map(_x_mult_2)
+        rs = MultiProcessingReadingService(num_workers=num_workers, worker_prefetch_cnt=worker_prefetch_cnt)
+        dl = DataLoader2(dp, reading_service=rs)
+        it = iter(dl)
+        for _ in range(EXCEPTION_ITERATION_NUM):
+            next(it)
+        with self.assertRaises(_CustomException) as cm:
+            next(it)
+        exc_msg = str(cm.exception)
+        self.assertTrue("Caught _CustomException in dispatching process" in exc_msg)
+        self.assertTrue("Original Traceback" in exc_msg)
+        self.assertTrue("_CustomException: oops" in exc_msg)
+
```

Comment:
nit: Maybe check if DL2 is shutdown properly? Is that possible to test?

Reply:
Actually, DL2 is not properly shutdown for now because we only raise Error in the main process. We haven't handled to shutdown other worker processes when on process has Error. This probably the same reason that it takes longer time to shutdown.

This is tracked in https://github.com/pytorch/data/issues/969